src/ext/php7/engine_hooks.c

```diff
@@ -547,3 +558,52 @@ void ddtrace_opcode_mshutdown(void) {
 
     zend_set_user_opcode_handler(ZEND_EXIT, NULL);
 }
+
+void ddtrace_execute_internal_minit(void) {
+    _prev_execute_internal = zend_execute_internal;
+    zend_execute_internal = _dd_execute_internal;
+}
+
+void ddtrace_execute_internal_mshutdown(void) { zend_execute_internal = _prev_execute_internal; }
+
+static void _dd_call_original_internal(zend_execute_data *execute_data, zval *return_value) {
```

Comment:
I recommend having another function pointer and point it at either the previous execute internal hook or execute_internal; no reason to keep checking that on every call.

Reply:
Brilliant!